File: barbarian.py
import random
from util import magic_weapon, do_roll, roll_dice
from character import Character
from feats import ASI, GreatWeaponMaster, PolearmMaster, AttackAction, Feat, EquipWeapon
from weapons import Glaive, Greatsword, GlaiveButt


class Beserker(Feat):
    def __init__(self, num_dice):
        self.name = "Berserker"
        self.used = False
        self.num_dice = num_dice

    def begin_turn(self, target):
        self.used = False

    def hit(self, args):
        if not self.used:
            self.used = True
            num = 2 * self.num_dice if args.crit else self.num_dice
            args.add_damage("Berserker", roll_dice(num, 6))


class BrutalStrike(Feat):
    def __init__(self, num_dice):
        self.name = "BrutalStrike"
        self.num_dice = num_dice

    def begin_turn(self, target):
        self.used = False

    def roll_attack(self, args):
        if not self.used and args.adv:
            args.adv = False
            self.enabled = True
            self.used = True

    def hit(self, args):
        if self.enabled:
            num = 2 * self.num_dice if args.crit else self.num_dice
            args.add_damage("BrutalStrike", roll_dice(num, 10))
        self.enabled = False

    def miss(self, args):
        self.enabled = False


class Retaliation(Feat):
    def __init__(self, weapon):
        self.name = "Retaliation"
        self.weapon = weapon

    def apply(self, character):
        self.character = character

    def enemy_turn(self, target):
        self.character.attack(target, weapon=self.weapon)


class PrimalChampion(Feat):
    def __init__(self):
        self.name = "PrimalChampion"

    def apply(self, character):
        character.str += 4


class Rage(Feat):
    def __init__(self, dmg):
        self.name = "Rage"
        self.raging = False
        self.dmg = dmg

    def apply(self, character):
        self.character = character

    def short_rest(self):
        self.raging = False

    def begin_turn(self, target):
        if not self.raging and self.character.use_bonus("rage"):
            self.raging = True

    def hit(self, args):
        if self.raging:
            args.add_damage("Rage", self.dmg)


class RecklessAttack(Feat):
    def __init__(self):
        self.name = "RecklessAttack"

    def begin_turn(self, target):
        self.enabled = True

    def roll_attack(self, args):
        if self.enabled:
            args.adv = True

    def end_turn(self, target):
        self.enabled = False


class Barbarian(Character):
    def __init__(self, level, use_pam=False):
        if level >= 16:
            rage_dmg = 4
        elif level >= 9:
            rage_dmg = 3
        else:
            rage_dmg = 2
        self.magic_weapon = magic_weapon(level)
        base_feats = []
        base_feats.append(Rage(rage_dmg))
        base_feats.append(RecklessAttack())
        if use_pam:
            weapon = Glaive(bonus=self.magic_weapon)
        else:
            weapon = Greatsword(bonus=self.magic_weapon)
        base_feats.append(EquipWeapon(weapon, savage_attacker=True))
        if level >= 5:
            attacks = 2 * [weapon]
        else:
            attacks = [weapon]
        base_feats.append(AttackAction(attacks=attacks))
        if level >= 3:
            base_feats.append(Beserker(rage_dmg))
        if level >= 17:
            base_feats.append(BrutalStrike(2))
        elif level >= 9:
            base_feats.append(BrutalStrike(1))
        if level >= 10:
            base_feats.append(Retaliation(weapon))
        if level >= 20:
            base_feats.append(PrimalChampion())
        if use_pam:
            feats = [
                GreatWeaponMaster(),
                PolearmMaster(GlaiveButt(bonus=self.magic_weapon)),
                ASI([["str", 1], ["con", 1]]),
                ASI(),
                ASI(),
            ]
        else:
            feats = [GreatWeaponMaster(), ASI([["str", 2]]), ASI(), ASI(), ASI()]
        super().init(
            level=level,
            stats=[17, 10, 10, 10, 10, 10],
            feats=feats,
            base_feats=base_feats,
        )

File: dndsim/src/classes/Barbarian.ts
import { Environment } from "../sim/Environment"
import { Operation } from "../sim/actions/Operation"
import { Character } from "../sim/Character"
import { Feat } from "../sim/Feat"
import { AttackRollEvent } from "../sim/events/AttackRollEvent"
import { WeaponMastery } from "../sim/types"
import { WeaponMasteries } from "../feats/shared/WeaponMasteries"
import { ClassLevel } from "../sim/coreFeats/ClassLevel"
import { applyFeatSchedule, defaultMagicBonus } from "../util/helpers"
import { ExtraAttack } from "../feats/shared/ExtraAttack"
import { Resource } from "../sim/resources/Resource"
import { Greatsword } from "../weapons"
import { SavageAttacker } from "../feats/origin/SavageAttacker"
import { GreatWeaponMaster } from "../feats/general/GreatWeaponMaster"
import { AbilityScoreImprovement } from "../feats/general/AbilityScoreImprovement"
import { IrresistibleOffense } from "../feats/epic/IrresistibleOffense"
import { SetAttribute } from "../feats/shared/SetAttribute"
import { IncreaseResource } from "../feats/shared/IncreaseResource"
import { AttackResultEvent } from "../sim/events/AttackResultEvent"

const RageResource = "rage"
const RageEffect = "raging"
const RecklessTag = "reckless"

class RageOperation implements Operation {
    repeatable = false

    eligible(_environment: Environment, character: Character): boolean {
        // While it is techically possible to rage while raging, at this time
        // there is no benefit to doing so
        return character.hasResource(RageResource)
            && character.bonus.has()
            && !character.hasEffect(RageEffect)
    }

    do(_environment: Environment, character: Character): void {
        character.useResource(RageResource)
        character.bonus.use()
        character.addEffect(RageEffect)
    }
}

const RageBonusDamageAttribute = "rageBonusDamage"

class Rage extends Feat {
    apply(character: Character) {
        this.addResource()
        character.customTurn.addOperation("before_action", new RageOperation())
        character.events.on("attack_result", (event) => this.attackResult(event))
        character.events.on("short_rest", () => this.shortRest())
    }

    attackResult(event: AttackResultEvent) {
        if (event.hit && event.attack.attack.stat(this.character) == "str") {
            event.addDamage({
                source: "Rage",
                flatDmg: this.character.getAttribute(RageBonusDamageAttribute),
            })
        }
    }

    shortRest() {
        this.character.removeEffect(RageEffect)
    }

    addResource() {
        this.character.resources.set(
            RageResource,
            new Resource({
                name: RageResource,
                character: this.character,
                initialMax: 2,
                incrementOnShortRest: true,
                resetOnLongRest: true,

            })
        )
    }
}

class RecklessAttack extends Feat {
    apply(character: Character) {
        character.events.on("attack_roll", (event) => this.attackRoll(character, event))
    }

    attackRoll(character: Character, event: AttackRollEvent) {
        if (event.attack.attack.stat(character) == "str") {
            event.adv = true
            event.attack.addTag(RecklessTag)
        }
    }
}

class PrimalChampion extends Feat {
    apply(character: Character) {
        character.increaseStatAndMax("str", 4)
        character.increaseStatAndMax("con", 4)
    }
}

class Frenzy extends Feat {
    used = false

    apply(character: Character) {
        character.events.on("begin_turn", () => this.beginTurn())
        character.events.on("attack_result", (event) => this.attackResult(event))
    }

    beginTurn() {
        this.used = false
    }

    attackResult(event: AttackResultEvent) {
        if (event.hit && !this.used && this.character.hasEffect(RageEffect) && event.attack?.hasTag(RecklessTag)) {
            const dice = Array(this.character.getAttribute(RageBonusDamageAttribute)).fill(6)
            event.addDamage({
                source: "Frenzy",
                dice,
            })
            this.used = true
        }
    }
}

class DivineFury extends Feat {
    used = false

    apply(character: Character) {
        character.events.on("begin_turn", () => this.beginTurn())
        character.events.on("attack_result", (event) => this.attackResult(event))
    }

    beginTurn() {
        this.used = false
    }

    attackResult(event: AttackResultEvent) {
        if (event.hit && !this.used && this.character.hasEffect(RageEffect) && event.attack?.attack.weapon()) {
            event.addDamage({
                source: "DivineFury",
                dice: [6],
                flatDmg: Math.floor(this.character.getClassLevel("Barbarian") / 2),
                // This could also be necrotic instead
                type: 'radiant',

            })
            this.used = true
        }
    }
}

export class Barbarian {
    static baseFeats(args: {
        level: number
        asis: Array<Feat>
        masteries: WeaponMastery[]
    }): Feat[] {
        const { level, asis, masteries } = args
        const feats: Feat[] = []
        if (level >= 1) {
            feats.push(new ClassLevel("Barbarian", level))
            feats.push(new WeaponMasteries(masteries))
            feats.push(new SetAttribute(RageBonusDamageAttribute, 2))
            feats.push(new Rage())
        }
        if (level >= 2) {
            feats.push(new RecklessAttack())
        }
        if (level >= 3) {
            feats.push(new IncreaseResource(RageResource))
        }
        if (level >= 5) {
            feats.push(new ExtraAttack(2))
        }
        if (level >= 6) {
            feats.push(new IncreaseResource(RageResource))
        }
        if (level >= 9) {
            feats.push(new SetAttribute(RageBonusDamageAttribute, 3))
        }
        if (level >= 12) {
            feats.push(new IncreaseResource(RageResource))
        }
        if (level >= 16) {
            feats.push(new SetAttribute(RageBonusDamageAttribute, 4))
        }
        if (level >= 17) {
            feats.push(new IncreaseResource(RageResource))
        }
        if (level >= 20) {
            feats.push(new PrimalChampion())
        }
        feats.push(
            ...applyFeatSchedule({
                newFeats: asis,
                schedule: [4, 8, 12, 16, 19],
                level,
            })
        )
        return feats
    }

    static berserkerFeats(level: number): Feat[] {
        const feats: Feat[] = []
        if (level >= 3) {
            feats.push(new Frenzy())
        }
        return feats
    }

    static zealotFeats(level: number): Feat[] {
        const feats: Feat[] = []
        if (level >= 3) {
            feats.push(new DivineFury())
        }
        return feats
    }

    static createBerserkerBarbarian(level: number): Character {
        const character = new Character({
            stats: { str: 17, dex: 10, con: 16, int: 10, wis: 10, cha: 10 },
        })
        const weapon = new Greatsword({ magicBonus: defaultMagicBonus(level) })

        const feats = [
            new SavageAttacker(),
            ...Barbarian.baseFeats({
                level,
                asis: [
                    new GreatWeaponMaster(weapon),
                    new AbilityScoreImprovement("str"),
                    new AbilityScoreImprovement("con"),
                    new AbilityScoreImprovement("con"),
                    new IrresistibleOffense("str"),
                ],
                masteries: ["Topple", "Graze"],
            }),
            ...this.berserkerFeats(level)
        ]
        feats.forEach((feat) => character.addFeat(feat))
        return character
    }

    static createZealotBarbarian(level: number): Character {
        const character = new Character({
            stats: { str: 17, dex: 10, con: 16, int: 10, wis: 10, cha: 10 },
        })
        const weapon = new Greatsword({ magicBonus: defaultMagicBonus(level) })

        const feats = [
            new SavageAttacker(),
            ...Barbarian.baseFeats({
                level,
                asis: [
                    new GreatWeaponMaster(weapon),
                    new AbilityScoreImprovement("str"),
                    new AbilityScoreImprovement("con"),
                    new AbilityScoreImprovement("con"),
                    new IrresistibleOffense("str"),
                ],
                masteries: ["Topple", "Graze"],
            }),
            ...this.zealotFeats(level)
        ]
        feats.forEach((feat) => character.addFeat(feat))
        return character
    }
}
